### PR TITLE
Branch ahollenbach

### PR DESCRIPTION
**Pull request by brandonshin**

### CODE REVIEW

**Comment by brandonshin** [comment]
Love it. Just merged it in!

File: slackbotExercise.py [hunk 1]
```diff
@@ -3,90 +3,280 @@
...
-    USERTOKENSTRING = settings['USERTOKENSTRING']
-    URLTOKENSTRING = settings["URLTOKENSTRING"]
-    TEAMNAMESTRING = settings["TEAMNAMESTRING"]
+    # The max number of users we are willing to look forward
+    # to try and find a good match
+    sliding_window = bot.sliding_window_size
 
+    # find a user to draw, priority going to first in
+    for i in range(len(bot.user_queue)):
+        user = bot.user_queue[i]
```
Comment (JulySept) [inline slackbotExercise.py:93]:
```suggestion
    for user in bot.user_queue:
```
I think it's better this way (:

File: slackbotExercise.py [hunk 2]
```diff
@@ -3,90 +3,280 @@
...
+        # User should be active and not have done exercise yet
+        if user in active_users and not user.hasDoneExercise(exercise):
+            bot.user_queue.remove(user)
+            return user
+        elif user in active_users:
+            # Decrease sliding window by one. Basically, we don't want to jump
+            # too far ahead in our queue
+            sliding_window -= 1
+            if sliding_window <= 0:
+                break
```
Comment (JulySept) [inline slackbotExercise.py:104]:
```suggestion
        if user not in active_users:
            continue
        # User should be active and not have done exercise yet
        if not user.hasDoneExercise(exercise):
            bot.user_queue.remove(user)
            return user
        else:
        # Decrease sliding window by one. Basically, we don't want to jump
        # too far ahead in our queue
            sliding_window -= 1
            if sliding_window <= 0:
                break
```

File: slackbotExercise.py [hunk 3]
```diff
@@ -3,90 +3,280 @@
...
 
-# Selects Next Time Interval and Returns the Exercise
-def selectExerciseAndStartTime():
+    for user_id in user_ids:
+        # Add user to the cache if not already
+        if user_id not in bot.user_cache:
+            bot.user_cache[user_id] = User(user_id)
+            if not bot.first_run:
+                # Push our new users near the front of the queue!
+                bot.user_queue.insert(2,bot.user_cache[user_id])
```
Comment (JulySept) [inline slackbotExercise.py:134]:
```suggestion
        if user_id in bot.user_cache:
            continue
        bot.user_cache[user_id] = User(user_id)
        if bot.first_run:
            continue
        # Push our new users near the front of the queue!
        bot.user_queue.insert(2,bot.user_cache[user_id])
```
this looks like less branches
(I'm just trying to do my homework, sorry)

File: slackbotExercise.py
Comment (JulySept) [inline slackbotExercise.py:1]: It may be worth introducing several constants (for example, for the delay time or for winner announcement)

**Pull request merged**